### DISCUSSION TITLE
Module graph outgoingConnections missing dependencies that are present in incomingConnections

### DISCUSSION THREAD

**Discussion by scameron** [webpack/webpack#18003] [General]
I'm looking at [ModuleGraph](https://github.com/webpack/webpack/blob/main/lib/ModuleGraph.js), specifically the `incomingConnections` and `outgoingConnections` structures.

Looking at the `outgoingConnections` for a particular, I noticed that there were no dependencies present for any of the dynamic `import`statements.  The dependencies from the static `import` statements are there, but not the dynamic ones.

Looking the other way around, at the `incomingConnections` for a module that was dynamically imported by the first module, I find the dependency is present.

Is this expected that `incomingConnections` contains dependencies that are not present in `outgoingConnections` ?  I was expecting to be able to get the same dependency info regardless of which structure I look at.  But that doesn't seem to be the case.  Is it correct?

Thanks in advance!

**Comment by scameron** [comment]
It looks like my problem is with concatenated modules.  If a module has been concatenated, then `compilation.moduleGraph.getOutgoingConnections` for that module is missing connections from the dynamic imports.  If I do `getOutgoingConnections` for the ConcatenatedModule itself, the dynamic import connections are there, but they are mixed in with all of the outgoing connections for any of the modules included in the concatenation.  I can't tell which are the connections for the specific module.

The incoming connections don't seem to have this problem.  If I get `compilation.moduleGraph.getIncomingConnections` for a module, it contains all incoming connections.

At the moment, to get the correct outgoing connection info, I need to iterate all of the incoming connection for all modules and invert those to figure out the corresponding outgoing connections.

This seems wrong, but maybe there's a reason for it.  Is it a bug or intended behavior?

**Reply by alexander-akait** [reply]
Sorry for the delay, yes it is expected, less memory and cpu usage, but if you have a case where you need it/have a problem, we can look into and discuss it.

**Reply by scameron** [reply]
I was able to determine the outgoing connections by starting with the incoming connections and inverting them.  That isn't too bad, although it makes our code a bit more complex because I need to precalculate the outgoing connections all at once.  It was more a matter of not knowing that outgoing connections doesn't have as much information in them as incoming connections, which was a bit surprising.
